evap/evaluation/models_logging.py

```diff
@@ -0,0 +1,286 @@
+            message = _("A {cls} was deleted.")
+
+        return message.format(
+            cls=self.content_type.model_class()._meta.verbose_name_raw,
+            obj=f'"{str(self.content_object)}"' if self.content_object else "",
+        )
+
+
+class LoggedModel(models.Model):
+    thread = threading.local()
+
+    class Meta:
+        abstract = True
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._logentry = None
+
+    def _as_dict(self):
+        """
+        Return a dict mapping field names to values saved in this instance.
+        Only include field names that are not to be ignored for logging and
+        that don't name m2m fields.
+        """
+        fields = [
+            field.name for field in type(self)._meta.get_fields() if
+            field.name not in self.unlogged_fields
+            and not field.many_to_many
+        ]
+        return model_to_dict(self, fields)
+
+    def _get_change_data(self, action_type: InstanceActionType):
+        """
+        Return a dict mapping field names to changes that happened in this model instance,
+        depending on the action that is being done to the instance.
+        """
+        self_dict = self._as_dict()
+        if action_type == InstanceActionType.CREATE:
+            changes = {
+                field_name: {FieldActionType.INSTANCE_CREATE: [created_value]}
+                for field_name, created_value in self_dict.items()
+                if created_value is not None
+            }
+        elif action_type == InstanceActionType.CHANGE:
+            old_dict = type(self).objects.get(pk=self.pk)._as_dict()
+            changes = {
+                field_name: {FieldActionType.VALUE_CHANGE: [old_value, self_dict[field_name]]}
+                for field_name, old_value in old_dict.items()
+                if old_value != self_dict[field_name]
+            }
+        elif action_type == InstanceActionType.DELETE:
+            old_dict = type(self).objects.get(pk=self.pk)._as_dict()
+            changes = {
+                field_name: {FieldActionType.INSTANCE_DELETE: [deleted_value]}
+                for field_name, deleted_value in old_dict.items()
+                if deleted_value is not None
+            }
+            # as the instance is being deleted, we also need to pull out all m2m values
+            m2m_field_names = [field.name for field in type(self)._meta.many_to_many if field.name not in self.unlogged_fields]
+            for field_name, related_objects in model_to_dict(self, m2m_field_names).items():
+                changes[field_name] = {FieldActionType.INSTANCE_DELETE: [obj.pk for obj in related_objects]}
+        else:
+            raise ValueError("Unknown action type: '{}'".format(action_type))
+
+        return changes
+
+    def _update_log(self, action_type: InstanceActionType, m2m_changes=None):
+        changes = self._get_change_data(action_type)
```

Comment:
it's still weird that when logging m2m changes, the code first starts comparing the instance with itself. can you find a nice way around that?

my suggestion would be to have

`def log_m2m_change(field_name, action_type, pk_set)`
`def log_instance_create/change/delete()`

and all of those call  `def _update_log(action_type, changes)`